core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java

```diff
@@ -0,0 +1,216 @@
+ * instance for a path. If this shows that your implementation is not getting discovered or used
+ * then you may wish to use the manual registration approach.
+ * <p/>
+ * <h4>Automatic Discovery</h4>
+ * <p>
+ * To use the {@linkplain ServiceLoader} based mechanism you need to have a file named
+ * {@code tachyon.underfs.UnderFileSystemFactory} placed in the {@code META-INF\services} directory
+ * of your project. This file should contain the full name of your factory types (one per line),
+ * your factory types must have a public unparameterised constructor available (see
+ * {@link ServiceLoader} for more detail on this). You can enable {@code DEBUG} level logging to see
+ * factories as they are discovered if you wish to check that your implementation gets discovered.
+ * </p>
+ * <p>
+ * Note that if you are bundling Tachyon plus your code in a shaded JAR using Maven make sure to use
+ * the {@code ServicesResourceTransformer} as otherwise your services file will override the core
+ * provided services file and leave the standard factories and under file system implementations
+ * unavailable.
+ * </p>
+ * <h4>Manual Registration</h4>
+ * <p>
+ * To manually register a factory simply pass an instance of your factory to the
+ * {@link #add(UnderFileSystemFactory)} method. This can be useful when your factory cannot be
+ * instantiated without arguments or in cases where automatic discovery does not give your factory
+ * priority. Factories registered this way will be registered at the start of the factories list so
+ * will have the first opportunity to indicate whether they support a requested path.
+ * </p>
+ */
+public class UnderFileSystemRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final List<UnderFileSystemFactory> FACTORIES =
+      new CopyOnWriteArrayList<UnderFileSystemFactory>();
+  private static boolean sInit = false;
+
+  static {
+    // Call the actual initializer which is a synchronized method for thread safety purposes
+    init();
+  }
+
+  static synchronized void init() {
+    if (sInit) {
+      return;
+    }
+
+    // Discover and register the available factories
+    ServiceLoader<UnderFileSystemFactory> discoverableFactories =
+        ServiceLoader.load(UnderFileSystemFactory.class);
+    Iterator<UnderFileSystemFactory> iter = discoverableFactories.iterator();
+    while (iter.hasNext()) {
+      UnderFileSystemFactory factory = iter.next();
+      LOG.debug("Discovered Under File System Factory implementation {} - {}", factory.getClass(),
+          factory.toString());
+      FACTORIES.add(factory);
+    }
+
+    sInit = true;
+  }
+
+  /**
+   * Resets the registry to its default state
+   * <p>
+   * This clears the registry as it stands and rediscovers the available factories.
+   * </p>
+   */
+  public static synchronized void reset() {
+    // Can't reset if we're not initialized
+    if (!sInit) {
```

Comment:
When would this be the case?  Currently all methods put a lock on the class object, so only one function can play around with mutating `FACTORIES`.

If there is a case where this would happen, wouldn't it be better to call init rather than returning?